File: yukicoder/tyama_yukicoder377.py
#!/usr/bin/python
from collections import defaultdict
from fractions import gcd
import sys,subprocess
if sys.version_info[0]>=3:
	raw_input=input
	xrange=range

def prime_division(n):
	h=defaultdict(int)
	line=subprocess.check_output(['factor',str(n)])
	for e in line.split()[1:]: h[int(e)]+=1
	return [(k,v) for k,v in h.items()]

def divisor_totient(a,d,n,t):
	if d==len(a):
		yield (n,t)
	else:
		for i in xrange(a[d][1]+1):
			for x,xt in divisor_totient(
				a,d+1,
				n*a[d][0]**i,
				(t if i==0 else t*(a[d][0]-1)*a[d][0]**(i-1))
			): yield (x,xt)

cache={}
M=1000000007
H,W,K=map(int,raw_input().split())
A=prime_division(H)
B=prime_division(W)
r=0
for a,at in divisor_totient(A,0,1,1):
	for b,bt in divisor_totient(B,0,1,1):
		key=W*H*gcd(a,b)//a//b
		if key not in cache: cache[key]=pow(K,key,M)
		r=(r+at*bt*cache[key])%M
print(r*pow(W*H,M-2,M)%M)

File: yukicoder/tyama_yukicoder377.rb
#!/usr/bin/ruby
def yield_from(*funcsyms)
	funcsyms.each{|funcsym|
		func = method(funcsym)
		define_method(funcsym){|*args,&blk|
			return to_enum(funcsym,*args) if !blk
			func.(*args){|*a|a.each{|e|blk.(e)}}
		}
		private funcsym
	}
end

if false
	require 'prime'
	def prime_division(n) n.prime_division end
else
	def prime_division(n)
		IO.popen('factor '+n.to_s){|io|
			a=[]
			line=io.gets
			line.split[1..-1].group_by{|e|e}.each{|k,v|a<<[k.to_i,v.size]}
			a
		}
	end
end
def divisor_totient(a,d,n,t)
	return to_enum(:divisor,a,d,n,t) if !block_given?
	if d==a.size
		yield [n,t]
	else
		(0..a[d][1]).each{|i|
			yield *divisor_totient(
				a,d+1,
				n*a[d][0]**i,
				i==0 ? t : t*(a[d][0]-1)*a[d][0]**(i-1)
			)
		}
	end
end
yield_from :divisor_totient
def pow(x,y,m)
	z=1
	while y>0
		z=z*x%m if y%2>0
		x=x*x%m
		y/=2
	end
	z
end

cache={}
M=1000000007
H,W,K=gets.split.map(&:to_i)
A=prime_division(H)
B=prime_division(W)
r=0
divisor_totient(A,0,1,1){|a,at|
	divisor_totient(B,0,1,1){|b,bt|
		key=W*H/a.lcm(b)
		cache[key]=pow(K,key,M) if !cache.has_key?(key)
		r=(r+at*bt*cache[key])%M
	}
}
p r*pow(W*H,M-2,M)%M
